interfaces/sourcegen/sourcegen/_types.py

```diff
@@ -0,0 +1,72 @@
+# This file is part of Cantera. See License.txt in the top-level directory or
+# at https://cantera.org/license.txt for license and copyright information.
+
+from abc import ABCMeta, abstractmethod
+from dataclasses import dataclass, fields
+import os
+
+
+def _unpack(obj):
+    return (getattr(obj, field.name) for field in fields(obj))
+
+
+@dataclass
+class Param:
+    """ Represents a function parameter """
+
+    p_type: str
+    name: str
+
+
+    @staticmethod
+    def parse(c_param: str):
+        parts = c_param.strip().rsplit(' ', 1)
+        if len(parts) == 2:
+            return Param(*parts)
+
+
+    def __iter__(self):
+        return _unpack(self)
```

Comment:
I like the `yield` statement! We have it in C# as well.

Reply:
OK looks like since the `__iter__` method needs a generator anyway, the `yield from` statement doesn't add much in this case. Plus I kinda liked the `_unpack` method because it is reused

Reply:
Ok, I've reworked this into decorator that will supply the `__iter__` method to the affected `@dataclass`es.